### PR TITLE
Admin dashbaord

### PR DESCRIPTION
**Pull request by Adimac93**
- cron reports update channel
- cron reports watch channel
- cron reports settings & stats API (unsecured)

TODO:
- secured settings & stats access (via `UserRole::Admin` check)
- store queue stats in aggregate DB
- store cron settings in DB
- frontend page

### CODE REVIEW

File: libs/block-mesh-manager/src/main.rs [hunk 1]
```diff
@@ -84,7 +85,14 @@ async fn run() -> anyhow::Result<()> {
...
     let ws_connection_manager = ConnectionManager::new();
-    let _reports_cron_task = ws_connection_manager.cron_reports(Duration::from_secs(60)); // FIXME
+    let _reports_cron_task = ws_connection_manager.cron_reports(
+        Duration::from_secs(60),
+        vec![
+            WsServerMessage::RequestUptimeReport,
+            WsServerMessage::RequestBandwidthReport,
+        ],
+        100,
+    );
```
Comment (Adimac93) [inline libs/block-mesh-manager/src/main.rs:95]: Expose inner cron settings.

File: libs/block-mesh-manager/src/ws/connection_manager.rs [hunk 2]
```diff
@@ -33,24 +34,88 @@ impl ConnectionManager {
...
-                            WsServerMessage::RequestUptimeReport,
-                            WsServerMessage::RequestBandwidthReport,
-                        ],
-                        100,
-                    )
-                    .await;
-                tokio::time::sleep(period).await;
-            }
-        })
+        let (tx, mut rx) = mpsc::channel::<CronReportSettings>(10);
```
Comment (Adimac93) [inline libs/block-mesh-manager/src/ws/connection_manager.rs]: Settings channel.

File: libs/block-mesh-manager/src/ws/connection_manager.rs [hunk 3]
```diff
@@ -110,14 +175,20 @@ impl Broadcaster {
         .await;
     }
 
-    pub async fn queue_multiple(&self, messages: &[WsServerMessage], count: usize) {
+    pub async fn queue_multiple(
+        &self,
+        messages: impl IntoIterator<Item = WsServerMessage> + Clone,
```
Comment (Adimac93) [inline libs/block-mesh-manager/src/ws/connection_manager.rs:241]: Accept more types that iter over `WsServerMessage` elements.

File: libs/block-mesh-manager/src/ws/connection_manager.rs [hunk 4]
```diff
@@ -37,79 +40,136 @@ impl ConnectionManager {
...
+            used_window_size,
+        }
+    }
 }
+impl Default for CronReportStats {
+    fn default() -> Self {
+        Self::new(vec![], 0, 0)
+    }
+}
+
```
Comment (Adimac93) [inline libs/block-mesh-manager/src/ws/connection_manager.rs:158]: Redundant

File: libs/block-mesh-manager/src/ws/connection_manager.rs [hunk 5]
```diff
@@ -241,103 +236,66 @@ impl Broadcaster {
...
-    println!("{stats:?}");
+async fn fetch_latest_cron_settings(
+    pool: &PgPool,
+    user_id: Uuid,
+) -> anyhow::Result<CronReportSettings> {
+    let aggregate =
+        get_or_create_aggregate_by_user_and_name_pool(pool, AggregateName::CronReports, user_id)
+            .await?;
+    let settings: CronReportSettings = serde_json::from_value(aggregate.value)?;
+    Ok(settings)
```
Comment (Adimac93) [inline libs/block-mesh-manager/src/ws/connection_manager.rs:300]: insert or update on table "aggregates" violates foreign key constraint "fk_user"

Comment (ohaddahan) [reply libs/block-mesh-manager/src/ws/connection_manager.rs:300]: No it doest if you use it as we said, it's not a random user but only a predefined "server user" , basically a constant.

File: libs/block-mesh-manager/src/ws/connection_manager.rs [hunk 6]
```diff
@@ -248,10 +252,28 @@ impl Broadcaster {
...
+            let user_id = Uuid::parse_str(
+                env::var(BLOCKMESH_SERVER_UUID_ENVAR)
+                    .context("Could not find SERVER_UUID env var")?
+                    .as_str(),
+            )
+            .context("SERVER_UUID evn var contains invalid UUID value")?;
+
             tokio::spawn(async move {
+                let _ = state
+                    .tx_aggregate_agg
```
Comment (ohaddahan) [inline libs/block-mesh-manager/src/ws/connection_manager.rs:265]: why you're using this?

File: libs/block-mesh-manager/src/ws/connection_manager.rs [hunk 7]
```diff
@@ -252,33 +254,49 @@ impl Broadcaster {
...
-            let pool = state.pool.clone();
+            let pool = pool.clone();
             let user_id = Uuid::parse_str(
                 env::var(BLOCKMESH_SERVER_UUID_ENVAR)
                     .context("Could not find SERVER_UUID env var")?
                     .as_str(),
             )
             .context("SERVER_UUID evn var contains invalid UUID value")?;
 
+            // TODO remove
```
Comment (ohaddahan) [inline libs/block-mesh-manager/src/ws/connection_manager.rs:265]: remove please

File: libs/block-mesh-manager/src/main.rs [hunk 8]
```diff
@@ -83,8 +84,18 @@ async fn run() -> anyhow::Result<()> {
...
+        .cron_reports(
+            Duration::from_secs(60),
+            vec![
+                WsServerMessage::RequestUptimeReport,
+                WsServerMessage::RequestBandwidthReport,
+            ],
+            100,
+            db_pool.clone(),
+        )
+        .await;
```
Comment (ohaddahan) [inline libs/block-mesh-manager/src/main.rs:98]: Strange to call some inner function that spawns a loop.
Should check if we can extract to a top level function that runs with some arguments like other background tasks.
With this structure, it's as if the ConnectionManager owns it.

File: libs/block-mesh-manager/src/routes/admin/reports_queue/change_settings.rs [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    State(state): State<Arc<AppState>>,
+    Extension(auth): Extension<AuthSession<Backend>>,
+    Json(body): Json<CronReportSettings>,
+) -> Result<impl IntoResponse, Error> {
+    let mut transaction = state.pool.begin().await?;
+    let user = auth.user.ok_or(Error::UserNotFound)?;
+    let user = get_user_opt_by_id(&mut transaction, &user.id)
+        .await?
+        .ok_or(Error::UserNotFound)?;
+    if !matches!(user.role, UserRole::Admin) {
```
Comment (ohaddahan) [inline libs/block-mesh-manager/src/routes/admin/reports_queue/change_settings.rs:32]: I changed the condition, this is more future proof, if we add another value to the enum in the future we don't need to remember to change this line.

File: libs/block-mesh-manager/src/routes/admin/reports_queue/get_stats.rs [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    let mut transaction = state.pool.begin().await?;
+    let user = auth.user.ok_or(Error::UserNotFound)?;
+    let user = get_user_opt_by_id(&mut transaction, &user.id)
+        .await?
+        .ok_or(Error::UserNotFound)?;
+    if !matches!(user.role, UserRole::Admin) {
+        return Err(Error::Unauthorized);
+    }
+    transaction.commit().await?;
+    if let Some(mut controller) = state
```
Comment (ohaddahan) [inline libs/block-mesh-manager/src/routes/admin/reports_queue/get_stats.rs:27]: Again, the controller owns this state which is weird.
The data is stored in the DB, and anyone that wants it simply access the DB.
The ConnectionManager shouldn't own it.

File: libs/block-mesh-manager/src/ws/broadcaster.rs [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+                stats_tx,
+            )
+            .await;
+        });
+
+        let controller = CronReportsController::new(stats_rx);
+        self.cron_reports_controller = Some(controller.clone());
+        Ok(controller)
+    }
+}
```
Comment (ohaddahan) [inline libs/block-mesh-manager/src/ws/broadcaster.rs:178]: I think we can simplify this a lot.
Don't think we need an extra channel.
This entire function seems redundant, only the loop should exist.

File: libs/block-mesh-manager/src/ws/connection_manager.rs [hunk 12]
```diff
@@ -33,124 +34,79 @@ impl ConnectionManager {
...
-                            WsServerMessage::RequestUptimeReport,
-                            WsServerMessage::RequestBandwidthReport,
-                        ],
-                        100,
-                    )
-                    .await;
-                tokio::time::sleep(period).await;
-            }
-        })
+    pub async fn cron_reports(
```
Comment (ohaddahan) [inline libs/block-mesh-manager/src/ws/connection_manager.rs:37]: A pass through function , can remove.

File: libs/block-mesh-manager/src/ws/connection_manager.rs [hunk 13]
```diff
@@ -33,124 +34,79 @@ impl ConnectionManager {
...
+    .await?;
+    let messages = messages.into();
+    loop {
+        let messages = messages.clone();
+        let settings = fetch_latest_cron_settings(pool, user_id).await?;
+        let new_period = settings.period.unwrap_or(period);
+        let new_messages = settings.messages.unwrap_or(messages);
+        let new_window_size = settings.window_size.unwrap_or(window_size);
+        let sent_messages_count = broadcaster
+            .queue_multiple(new_messages.clone(), window_size)
```
[the remark this answers is not in the document]
Comment (ohaddahan) [reply libs/block-mesh-manager/src/ws/connection_manager.rs:82]: ok i see the naming is wrong (i changed it)
this should be called something like "connect_ws_nodes_dispatcher" or something better.

File: libs/block-mesh-manager/src/ws/connection_manager.rs [hunk 14]
```diff
@@ -33,124 +34,79 @@ impl ConnectionManager {
...
-            sockets: Arc::new(DashMap::new()),
-            queue: Arc::new(Mutex::new(VecDeque::new())),
+pub async fn settings_loop(
+    pool: &PgPool,
+    user_id: &Uuid,
+    period: Duration,
+    messages: impl Into<Vec<WsServerMessage>> + Clone + Send + 'static,
+    window_size: usize,
+    broadcaster: Broadcaster,
+    stats_tx: Sender<CronReportStats>,
```
Comment (ohaddahan) [inline libs/block-mesh-manager/src/ws/connection_manager.rs:57]: Someone is listening on the RX side? It seems to be in main.rs and the value isn't used.

File: libs/block-mesh-manager/src/routes/admin/reports_queue/get_stats.rs (hunk 10 above)
Comment (Adimac93) [reply libs/block-mesh-manager/src/routes/admin/reports_queue/get_stats.rs:27]: Are you saying that we should simply keep the stas in the database in place of locally owned channels?

File: libs/block-mesh-manager/src/ws/broadcaster.rs (hunk 11 above)
Comment (Adimac93) [reply libs/block-mesh-manager/src/ws/broadcaster.rs:178]: Yes, it's possible if we keep stats in the DB.

**Pull request merged**